### ISSUE TITLE
During `fisher update`, init script is executed multiple times

### ISSUE THREAD

**Issue by mnewt** [fisherman/fisherman#397]
If there are multiple plugins installed, each init script is called multiple times. The number of times is related to the number of plugins installed. Environment variables are only set during the last call.

What's going on here?

Here is a simple reproduction:
```
> fisher up
OK fisherman is up to date
Updating 3 plugins
OK Latest fix
Backup fisher.fish
Synced 2 symlink/s
Done in 2s 946ms
> ls
init.fish  test.fish
> cat init.fish
echo init! path: $path
> cat test.fish
echo test! path: $path
> fisher .
Installing ~/a-plugin
init! path: /Users/mnewt/.config/fisherman/a-plugin
Done in 93ms
> fisher up
init! path:
OK fisherman is up to date
Updating 4 plugins
init! path:
OK Latest fix
init! path: /Users/mnewt/.config/fisherman/a-plugin
test! path: /Users/mnewt/.config/fisherman/a-plugin
Backup fisher.fish
Synced 3 symlink/s
Done in 2s 422ms
```

Cheers

**Comment by jbucaran** [comment]
@mnewt I'm looking at this, but in the mean time, can you tell me why are you using init.fish files?

If you can avoid them, please do, they are not a fish-shell thing and you should be using [conf.d](https://fishshell.com/docs/current/index.html#initialization) instead.

**Comment by mnewt** [comment]
@jbucaran `init.fish` is from oh-my-fish.
https://github.com/oh-my-fish/oh-my-fish/blob/master/docs/en-US/Packages.md#init

After reading about `conf.d`, it is cleaner. However, I'm trying to make my plugin compatible with both systems.

**Comment by jbucaran** [comment]
@mnewt Yep, I know about `init` as I was who introduced it before conf.d was even a thing, but now I regret it. 😅 

Can you show me the source code of the init.fish file you are running, that could help find out what's going on. I am hoping this is an issue in that file and not with fisherman.

**Comment by mnewt** [comment]
The whole source code is in the first post:
`echo init! path: $path`

**Comment by mnewt** [comment]
It's not unique to `init.fish`. The same thing happens when a `.fish` file is placed in `conf.d`.

**Comment by jbucaran** [comment]
@mnewt Okay, I had a chance to try this and here is what how things look on my end.

<img width="722" alt="screen shot 2017-09-02 at 12 34 33" src="https://user-images.githubusercontent.com/56996/29992616-27660712-8fdb-11e7-852b-9c54ceb61031.png">

**Comment by jbucaran** [comment]
@mnewt Oh, okay, I was able to repro the problem and now I know what you mean.

**Comment by jbucaran** [comment]
@mnewt So, I remember now. The reason we ignore this is because init.fish or conf.d files rarely put any code that prints to the screen, so we tend to forget that initialization files init.fish/conf.d run multiple times during fetch / updates.

The reason is fisherman's concurrent fetching. Whenever we install or update one or more plugins, a  new fish-shell process is created and any initialization files will run inside that process as usual.

**Comment by mnewt** [comment]
@jbucaran Ah, that makes complete sense. It is a bit messy though.

I'm thinking out loud, having skimmed the code but not understanding everything, I have two questions:
1. Just to confirm: Why is `$path` variable not set, except for on the last run? Is it because `fish` is invoking the script directly, as opposed to `fisherman` running it?
2. Would it be a big deal to re-work things such that the fetching is in parallel but the linking happens at the end? If something like this seems a good solution to you, I might be willing to take a crack at it.

**Comment by jbucaran** [comment]
1) Yes, it's fish that's running conf.d (as well init.fish which fisherman translates into correct conf.d).
2) That's possible, but it'll get messier :)

**Comment by mnewt** [comment]
@jorgebucaran It looks great to me. I completely agree with your KISS approach. Many thanks for making an awesome piece of software even awesome-er!

**Comment by jorgebucaran** [comment]
I'm glad you say so. Thank you, @mnewt. 🙇
